### ISSUE TITLE
Improved Platformer Example

### ISSUE THREAD

**Issue by charburgx** [GDevelopApp/GDevelop-examples#223]
## Describe the example

Improved platformer example that aligns with the website design.

_Note: all sounds used are adapted from CC0 sounds or made by me. Here's a list of used sounds:_

- https://freesound.org/people/taure/sounds/575320/
- https://freesound.org/people/Elektronikz/sounds/244369/
- https://freesound.org/people/felix.blume/sounds/328293/

## Checklist

- [x] I've followed all of [the best practices](http://wiki.compilgames.net/doku.php/gdevelop5/community/guide-for-submitting-an-example).
- [x] My game has a proper name in the game properties. 
- [x] My game package name behind with `com.example.`.
- [x] My game has all events unfolded.
- [x] I've added myself as the author in the game properties.
- [x] I've included a file called "README.md" with a description in proper English, explaining what this example is doing.
- [x] I confirm that this game and all of its resources can be integrated to this GitHub repository, distributed and MIT licensed.
- [x] I've cleaned unused resources.

## Game folder

[platformer_ver_1_0_0.zip](https://github.com/GDevelopApp/GDevelop-examples/files/8064621/platformer_ver_1_0_0.zip)

**Comment by 4ian** [comment]
Looking *great*! Great start. Careful I'll be having a lot of comments because this will be becoming the first and most opened game example soon, so it must be perfect ;)

**Asset size**:

A few things that I noted, which are basically all optimisations (because you can tell this was adapted from something made by designers, not game artists ;)):

- The game resolution is quite large (1920x1080), this gives a lot of work for mobile (or even desktop) devices, I think we could go for something a bit smaller, like 960x540 (i.e: divide it by two). 
- Similarly, the assets are very large, which creates performance problem, big download size and the background can't be displayed on mobile because the GPUs can't handle an image this big. 
  - For example, platform4.png is 1500x1287, while it's "just" a platform. This gives tons of work for the GPU. Reduce this by 3 to 500x429. Same for almost all other assets, **let's divide the size of almost all assets by 3**.
    - The rule of thumb is that once you have chosen the size of the red character (his size is good), then all other assets should be proportional to it. You should not need to resize platforms or anything in the scene editor.
  - 180 images for the fly is *waaaay* too much. We should reduce this to 10 at most (this will reduce the quality of the animation, but it's a game and we can't afford 180 images for a single enemy).
  - Also try to crop as much assets as possible so that they are not having transparent margins around them (this costs memory).
  - For backgrounds, reduce them (they are not displaying on mobile because larger than 2048 pixels) and use a tiled sprite for the space background.
  - Finally, the mp3 are also huge, one is 7mb. We should reduce them to 100-200kb max. We might need to either reduce their quality or cut them.

This is arguably a problem that the app is not warning you about this, we have something on the roadmap to display warnings but we never implemented these.

**Gameplay**:

I would maybe add an effect or two when destroying the second enemy, and make the level a little longer. 

**Events and behaviors**:

In terms of implementation, I think we should try to use more behaviors to reduce the size of the events sheet:

- let's try to use "Parallax for Tiled Sprite" behavior for the foreground and backgorund.
- Unfold all event groups (we don't fold anything for examples)
- Avoid group names like "Init", this is frightening for a non technical person :) 
- Let's us a smooth camera extension for following the player.
- You don't need the extra variable here:
  <img width="846" alt="image" src="https://user-images.githubusercontent.com/1280130/154047605-8dbc1422-02e8-427b-991f-eed1d301a0ef.png">
  
Basically we must try to find a good balance between using enough reusable behaviors, and still keep things in the events sheet but without overcomplicating something.
In case of doubt between: should I do something simple with a behavior or should I do exactly what I want with 3 or 4 events, always choose the behavior.

I think we'll get a great example this way! :)

**Comment by charburgx** [comment]
[Platformer_ver1_0_1.zip](https://github.com/GDevelopApp/GDevelop-examples/files/8086392/Platformer_ver1_0_1.zip)

Note: In this version I introduced two new extensions. One is "SimpleSmoothCameraFollow" and the other is "VolumeFalloff." The goal was to keep smooth camera following and volume falloff but without using lerp or complicated events directly.

As for the camera extension, I see that Davy has a PR for something like this, but if I understand correctly, it is a relatively advanced extension which he needs time to work on. I propose that in the meantime, we release an extension like the one in this file, which is essentially just a drop-in replacement for the "Center the camera on a player within bounds" action from the standard actions, but smooth.

I propose we then do the same for the "Volume Falloff" extension.

If you're okay with this, then I'll go ahead and complete the metadata for both of these extensions and make PRs for them to be part of the official extensions. Then I can add them into this project.

Otherwise, we can consider restricting the functionality of the game so we don't need these extensions, and we can also consider waiting for Davy to finish his camera following extension!

Anyway, here's a changelog:

- Game resolution in half
- All animations reduced to max 10 frames
- All image assets resized to their in-game size and padding removed
- All audio assets cut and compressed 112 kbps mp3 (the largest, crickets.mp3, is now ~150kb)
- Everything that can be a tiled sprite is now a tiled sprite
- Level made a little longer
- Added particle effect on killing the second monster
- Fixed last platform clipped (Wesley mentioned this in Discord)
- Fixed mobile control touchscreen issue (Wesley also mentioned this in Discord)
- Removed door glow having sinusoidal opacity over time
  - If you want to re-include this, we could make another extension, something like "Oscillating Variable" which gives a variable the value of a sin function with variable amplitude/period
 - Removed checkpoint glow due to significant clutter in the event sheet
 - Renamed "Init" group to "Beginning"
 - Replaced manual Parallax with the relevant extension
 - Improved collision boxes
 - Replaced smooth camera logic with new "SimpleSmoothCameraFollow" extension
 - Replaced volume falloff logic with new "VolumeFalloff" extension

There's probably more, but that should be most of it.

**Comment by 4ian** [comment]
Great improvements so far! And thanks Bouh for your review. A few more stuff from my side ;)

* Gameplay: the collision masks are not exactly rectangles, so the player can walk on the edges:
  <img width="968" alt="image" src="https://user-images.githubusercontent.com/1280130/154529128-cab1ed12-2c6f-45a5-95ba-83f906eab4aa.png">
  <img width="677" alt="image" src="https://user-images.githubusercontent.com/1280130/154528945-421da63d-c8e1-4384-ad0f-1f5c0e53a4a3.png">

  This is arguably because of the UI which is not helpful... but still go through the collision masks and make sure all X/Y positions are exactly aligned for all edges (take the opportunity to use round numbers).

* Sound design: I love the background musics and jump sound and the sounds effects 👏
* Level design: I still feel like we could add a platform at the top of the ladder, with a "jump tru" platform behavior, so that when we reach the top we stop on the platform :) 
* For events:
  * Still some events that can be behaviors: you should be able to replace this:
    <img width="1111" alt="image" src="https://user-images.githubusercontent.com/1280130/154530142-5a10f008-2aec-4d31-bbea-43d9fce75164.png">

    by either Sine Movement, Rectangular Movement or maybe Timed Back and Forth Movement (but without flipping the sprite). We'll have 4 less events 👍
  * Nitpicking: Collision condition are fairly expensive, prefer to run once and do sub events according to if the player is falling or not:
  <img width="916" alt="image" src="https://user-images.githubusercontent.com/1280130/154529764-2393ac55-1aa4-4f40-a83f-8e6c06d3f5e6.png">

  Same for the fly - this will also make the events a tiny bit easier to read (less repetition)

With these additional behaviors and rework, I think we should get to something fairly approachable and simple :)

**Comment by charburgx** [comment]
[Platformer_v_1_0_2.zip](https://github.com/GDevelopApp/GDevelop-examples/files/8093127/Platformer_v_1_0_2.zip)

> Level design: I still feel like we could add a platform at the top of the ladder, with a "jump tru" platform behavior, so that when we reach the top we stop on the platform :)

I don't have a specific sprite for this. Do you think it's best to make a wooden platform that looks similar to the ladder sprite?

If you think it should be a similar looking platform to the others, this poses a potential problem since it would require duplication of one of the platform objects (one that is jump-thru, one that is not). 

Let me know how you feel on this :)

> On the ladder, the hero doesn't have a back animation? It looks strange actually to up on a side.

Yes, I agree, I simply tried the best with the assets we have.

Ideally, we would have a specific climbing animation. I am happy to try making it myself, however since I have 0 experience with character animation, I cannot guarantee that the end result would look excellent.

> Door haven't interaction, I think we can add a message on the screen when you use it, like "Congratulation you won XXX coins!", something to understand that we've finished playing the level.

That would be nice, but I'm not sure how exactly to implement this simply and without introducing significantly more events. Do you have an exact idea on how this could look and still be implemented simply?

> Gold diamond look like a big coin not a checkpoint, but why not

I am simply copying the design on the gdevelop.io homepage. I figure the designer was going for an element of visual consistency by keeping them the same color - what kind of design do you think we might go with for the checkpoint instead?

> The tiled sprite object BackgroundPlants isn't really tiled, we can see the borders.

I don't have a great idea for a solution to this. We can include a flipped version of the background sprite, but this will double the file size. Do you think its noticeable enough to be worth remaking the background?

Other than this, here's the changes in the linked version:

 - Thumbnail.png reduced to 100px
 - Extensions changed to comply with extension guidelines
 - [Camera extension posted](https://github.com/GDevelopApp/GDevelop-extensions/pull/372)
 - [Volume extension posted](https://github.com/GDevelopApp/GDevelop-extensions/pull/374)
 - Last enemy moved to the right to make it easier to get on the platformer
 - Removed unknown behavior from player object
 - Replaced camera bound
 - Collision masks fixed
 - Replaced the movement with Rectangular Movement with "vertical distance" set to 0, since the two "move back and forth" extensions flip the sprite, and I couldn't figure out a way to change that behavior
 - Grouped the collisions for greater simplicity in events/better performance

Let me know if I missed anything !!

**Comment by charburgx** [comment]
[Platformer_v_1_1_0.zip](https://github.com/GDevelopApp/GDevelop-examples/files/8139642/Platformer_v_1_1_0.zip)

Thanks for all the great feedback! I think this version is significantly better :)

> Light object for door glow

I ended up not doing this for a few reasons: for one, it makes things quite complicated for a simple started project. Also, the light shadows do not play well with the hitboxes of the platforms. Plus I'm not sure of a way to light the foreground separately from the background.

> The background tiled object's offset could be changed to achieve the movement effect like on the old platformer.

This is the case currently, at least if you're referring to the parallax. It's subtle, but I am confident that subtle is best for this, since we want the foreground to feel separate from the background to reduce the feeling of 'clutter.'

Changes:

- Created climbing animation
- Adjusted the size of most objects to be smaller in relation to the player (Once we all are in agreement about the size, I will post a final version to resize the files to decrease sizes)
- Shifted the background down to decrease clutter
- Scaled cloud sizes by 2/3 to decrease clutter
- Re-rasterized the enemy assets to attain maximum quality
- Added invisible jumpthru platform to top of ladder
- Added end screen upon interacting with door
- Changed checkpoint to eye sprite
- Adjusted background y offset so that the tiling is significantly less obvious
- Add walls to left and right side of game window
- Disabled constant movement on platform
- Added vertical motion to the fly
- Cleaned up events with grouping
- Removed unused assets
- Removed health extension, replaced with boolean variables
- Removed .DS_Store files (I think, OS X is annoying about this)
- Removed markdown header from README

**Comment by 4ian** [comment]
Thanks everyone for your feedback!

I like the new ending screen and the game is nicer to play now. Still a few stuff on my side 😉

Gameplay:

* Reduce the Monster Fire collision mask to a smaller rectangle. It's super irritating to be considered to touch the monster while you did not because of the transparent margins around an object. Typically the kind of thing that make you rage qui the game and GDevelop. Always make things in favor of the player: to be considered to touch the fire, you must really be colliding with it, and even add a safety margin so you can be seen on top of it and still the game gives you a pass.
* Let's reduce again the platforms sizes. Even in the latest version it still gives the impression the player is not super agile or won't make it during a jump because platforms are huuuge. Again it's because we based this on assets that were made to be beautiful, but not made to be interesting in terms of gameplay. Still, I think it's worth making the platform smaller.
  Also this will allow to move platforms to create more interesting layout. The moving platform is taking one third of the screen height, that's huuuuuge.
  
Events:
  
* Can we rework these events to be using an object timer and a an object variable? This is giving otherwise a bad example that scene variables are the way to handle multiple objects, while every object should work on its own: 
  <img width="1089" alt="image" src="https://user-images.githubusercontent.com/1280130/155729614-7d9b9ec8-d2e3-4f47-b88c-98642ec547bd.png">
   These events are also a bit complicated I think, or maybe an animation would have worked better?

* The "Grass Sounds" events are taking a huge space (basically a small screen height) for something not that important for a new user. Should we find a way to hide them? Or do something simpler.


Nitpickings:
* ScoreUI => Score (or ScoreText). For a new user, UI is not a well known acryonym (and it's not a user interface, it's a text).
* Remove the toggled off event.
* There is "an unknown or unsupported instruction".
* Unknown behaviors on objects to remove.
* Is the Fly object removed when out of screen? It's important we clean up everything.
* Let's add a second monster and a third fly (even if it's making the level a bit longer, but with smaller platforms not necessarily). Important to show that all objects are working even when there are more than one.
* <img width="1100" alt="image" src="https://user-images.githubusercontent.com/1280130/155729204-4ca4cfc7-c637-4fe6-8189-d5f7c6309a35.png">
  This is frightening for a new user. Can we simplify the first event or even remove it? I'm willing to have stuff a bit less polished if it simplifies the events.
  
About other remarks:

> Use a light object for the door glow. Maybe light will be too complex for new users?

Let's not to it to keep the example simple and efficient.

> The example shouldn't use the Health extension because I think it makes the things a bit too abstract for something really simple

I have fixed feeling on this but agree this can be done with a simple variable so let's avoid Health then.

**Comment by tristanbob** [comment]
> * The "Grass Sounds" events are taking a huge space (basically a small screen height) for something not that important for a new user. Should we find a way to hide them? Or do something simpler.

I think we should use a single audio file (most users won't have multiple sound files) and randomize the speed it plays, like you show in this example:  https://www.youtube.com/watch?v=3akue2EwDPU

https://user-images.githubusercontent.com/8879811/155894725-d877f49d-0218-47c5-bddc-f6b811646a69.mp4

**Comment by 4ian** [comment]
All good stuff! 
A bit sad about the parallax extension, I know it's making us save only one event which is simple, but in the goal of promoting more games built on behaviors, I would still have kept it... but I'll leave it up to you to decide.

I think I had a bug where the running footsteps sounds and effects were not launched. I think this can be 100% fixed with the condition checking if an object is on specific floor:
<img width="383" alt="image" src="https://user-images.githubusercontent.com/1280130/156164454-9aa2f4ad-745d-4586-acaf-410a654883bc.png">

Because this is the floor condition from the platformer object, it's guaranteed to work 100% of the time if the player is on a floor. Otherwise, with the collision condition, there is a small risk there is a gap between the player and the platform (can happen).

Apart from that, and the bug to investigate, **I'm happy with the state of the game**! Maybe a final addition would be to demonstrate putting a button on the end layer, and reset the scene when played. If it can be done in a few events, as simply as possible, then it's good to do it. 

Great work @charburgx, Wesley and everyone there :) Let's go ahead with a final version :)

**Comment by charburgx** [comment]
[Platformer_v_1_1_5.zip](https://github.com/GDevelopApp/GDevelop-examples/files/8172334/Platformer_v_1_1_5.zip)

Alright, here it is! I added the floor test condition as well as the retry button. I also minified a number of the assets to keep the bundle size to a minimum 👍 

Go ahead and have one last look for any last comments :)

Thanks to everyone who gave feedback! I think this is gonna be a great addition to the game.

**Comment by Bouh** [comment]
Amazing!

Here a short things I noticed:

- [ ] Player can't go down on the ladder, think to check inside the player behavior this checkbox: 
![image](https://user-images.githubusercontent.com/1670670/156759686-5ca7ee4a-3f4f-4242-9ac8-c3d019b05ffa.png)
- [ ] I suggest FPS max to 0 for unlimited ?
- [ ] Check the first toggle to get same collision mask on all animations, specially for climbing animation you added (btw sprite is nice when climbing! ) Otherwise the Hero is a gym monster 
![image](https://user-images.githubusercontent.com/1670670/156762256-48f31883-fe9a-4363-8a85-6a9be309764c.png)

- [ ] Mobile inputs doesn't works because of this main event block the sub-events:
![image](https://user-images.githubusercontent.com/1670670/156761395-d3b2595c-7ef4-4a63-a2c1-b208ed5110a4.png)
- [ ] crickets.ogg is terrible on smartphone, there is a continious noise behind the crickets, it's aweful on smartphone
- [ ] Not sure we want show to people missing values 
![image](https://user-images.githubusercontent.com/1670670/156762029-6bc1ebea-194a-4ab1-b3f0-81261f90ffb8.png)

**Comment by 4ian** [comment]
> I suggest FPS max to 0 for unlimited ?

No it's a good idea to keep it at 60 especially for mobile devices to avoid draining their batteries by rendering too much.

**Comment by charburgx** [comment]
[Platformer_v_1_1_6.zip](https://github.com/GDevelopApp/GDevelop-examples/files/8199994/Platformer_v_1_1_6.zip)

> crickets.ogg is terrible on smartphone, there is a continious noise behind the crickets, it's aweful on smartphone

I tested this on my Google Pixel 2 headphones, both on the web build and apk build, and was unable to repeat this issue. It may be a problem with your phone speakers?

Changes:
 - Allow player to go through jumpthru platforms
 - Toggled same collision mask for all player animations
 - Made ladder collision box thinner (to prevent player from hanging off of the side)
 - Removed missing values from smooth camera function
 - Fixed mobile controls

**Comment by D8H** [comment]
I think the camera following extension should be a behavior. This way, the video will still be up to date when the extension is released.

Also, there should be an independent action to keep the camera in the area as the action will either be added by this PR:
* https://github.com/4ian/GDevelop/pull/3703
or added in a community extension if the PR is rejected.

**Comment by zatsme** [comment]
This is nice 😃

Can I suggest adding vertical movement to the parrllax and camera shake would be nice landing from a fall?

**Comment by charburgx** [comment]
[Platformer_v_1_1_7.zip](https://github.com/GDevelopApp/GDevelop-examples/files/8219120/Platformer_v_1_1_7.zip)

Made a new behavior for checkpoints as per your request @4ian. If there is no immediate feedback for it I'll make a post for it on the extensions repo :)

> Camera shake

Would be very nice! We're trying to limit the amount of events as much as possible so I haven't added it, but if others think that the positive of having it in outweighs that then it will be added for sure.

> Vertical movement to the parallax

Since we decided to go with camera offset, this would probably add a little more complexity than we would like. If we decide to go back to using the parallax extension then this would be more than doable.

Changes:

 - Made controls significantly larger
 - Moved checkpoint actions to new "Simple Checkpoint" behavior

**Comment by D8H** [comment]
Setting an effect might calculate a new image for the object, it should be done only once.
Can a checkpoint be disactivated? I don't see when it should and it stays activated in the preview.

This event seems to work the same:

![image](https://user-images.githubusercontent.com/2611977/158772456-70971899-56f3-4afd-b8a5-9ae64cdeabec.png)

About "Active a checkpoint" parameters:
- "Set spawn point", in which use case could it be set to false?
- "Set spawn point even if checkpoint is already activated", this should probably be a property as it probably won't change during a game.

**Comment by Bouh** [comment]
@charburgx how are you getting on with changes?
Need a new review?

**Comment by Bouh** [comment]
I made tiny changes:
- Rework what suggested by @D8H in [previous comment just above](https://github.com/GDevelopApp/GDevelop-examples/issues/223#issuecomment-1070551490).
- Add trigger once on actions "Key pressed left/right" to flip the object once, and not every frame keys are pressed. (a tiny optimization)

Nothing more changes, it have to fit the first video.
